test/tests/dom/shadowDOMFunctions.js

```diff
@@ -0,0 +1,148 @@
+    }
+
+});
+
+describe('isShadowElement cases', () => {
+  
+    it('should return true if parent node is shadow root',  () => {
+        const innerElement = document.querySelector('custom-web-component')?.shadowRoot?.querySelector('#inner-span');
+
+        if (!innerElement) {
+            throw new Error('there is not inner element');
+        }
+        
+        const result = isShadowElement(innerElement);
+
+        if (!result) {
+            throw new Error(`Expected result to be true, got ${ String(result) }`);
+        }
+    });
+
+    it('should return false if parent node is not shadow root', () => {
+        const testElement = document.createElement('div');
+        const result = isShadowElement(testElement);
+
+        if (result) {
+            throw new Error(`Expected result to be false, got ${ String(result) }`);
+        }
+    });
+});
+
+describe('getShadowRoot cases', () => {
+    it('should return shadow root host', () => {
+        const innerElement = document.querySelector('custom-web-component')?.shadowRoot?.querySelector('#inner-span');
+
+        if (!innerElement) {
+            throw new Error('there is not inner element');
+        }
+        const result = getShadowRoot(innerElement);
+
+        if (!result) {
+            throw new Error(`should have returned innerElement`);
+        }
+
+        if (!result.toString() === '[object ShadowRoot]') {
+            throw new Error(`should have returned '[object ShadowRoot]', gotten ${ result.toString() }`);
+        }
+        
+    });
+});
+
+describe('getShadowHost cases', () => {
+    it('should return shadow host if exists', () => {
+        const innerElement = document.querySelector('custom-web-component')?.shadowRoot?.querySelector('#inner-span');
+
+        if (!innerElement) {
+            throw new Error('there is not inner element');
+        }
+
+        const result = getShadowHost(innerElement);
+        
+        if (!result) {
+            throw new Error(`should have returned the inner element, but gotten undefined`);
+        }
+
+        const hostId = result.getAttribute('id');
+
+        if (hostId && hostId !== 'shadow-host') {
+            throw new Error(`should have returned> shadow-host but gotten ${ hostId }`);
+        }
+
+    });
+});
+
+describe('insertShadowSlot cases', () => {
+    it('should throw exception if element is not in shadow DOM', () => {
+        const testElement = document.createElement('div');
+
+        try {
+            insertShadowSlot(testElement);
+        } catch (error) {
+            console.log(error.message);
+            if (!error.message.match(/Element is not in shadow dom/i)) {
+                throw new Error(`should have thrown 'Element is not in shadow dom' exception, gotten: ${ error.message }`);
+            }
+        }
+
+    });
+
+    // don't know how to test it
+    it.skip('should throw exception if Host element is also in shadow dom', () => {
```

Comment:
I think we need another web component to be able to test this use case. Ex: `customElements.define('custom-web-component-wrapper' ...` Can you try this out?

Reply:
Thanks for the suggestion! and just for record, it was no as trivial as doing this:

```html
<html>
      ...
     <wrapper-custom-component>
             <inner-component>
                    <div id="inner-div">
                          Text from a nested custom component
                    <div>
             </inner-component>
     </wrapper-custom-component>
    ...
</html>
```

What I did for testing the failing case (when shadoRoot is also shadowHost in `insertShadowSlot()` function) was to attach the  ` <inner-component>` to a new div designed as a host within the `<wrapper-custom-component>` and I could reproduce the expection`Host element is also in shadow dom`